woocommerce/woocommerce
# Package release script: Prepare packages

### All Submissions:






### Changes proposed in this Pull Request:

This PR automates changelog creation and version bumping to prepare JS packages before release by creating a Package Release tool. It will perform the task on one, several, or all JS packages and skips private packages or ones Solaris is responsible for. The next step will be to create a pull request before making the actual release.

Part of https://github.com/woocommerce/woocommerce/issues/33319

### How to test the changes in this Pull Request:

1. `pnpm install`
2. Run the prepare tool on `@woocommerce/components` which has changelog entries ready to go (as of this PR being created).
```
./tools/package-release/bin/dev prepare @woocommerce/components
```
3. Check the diff to see that the changelog was appropriately created, the version was correctly bumped in `package.json`, and no errors resulted.
4. Stash changes and try on a package with no changelog entries. No changes should result.
```
./tools/package-release/bin/dev prepare @woocommerce/number
```
5. Try preparing a package that does not exist. An appropriate error should result.
```
./tools/package-release/bin/dev prepare @woocommerce/penguin-team 
```
6. Try preparing a private package. An appropriate error should result.
```
./tools/package-release/bin/dev prepare @woocommerce/notices 
```
7. Try releasing all packages and check the diff to make sure its ok.
```
./tools/package-release/bin/dev prepare -a
```
8. Try supplying no arguments. An appropriate error should result.
```
./tools/package-release/bin/dev prepare
```
> **Note**: Previous changelog entries have a formatting problem where `-` are re-added when previous entries are formatted. This will be handled separately. This is being fixed in https://github.com/woocommerce/woocommerce/pull/33527

```md
## [10.1.0](https://www.npmjs.com/package/@woocommerce/components/v/10.1.0) - 2022-06-09 

-    - Add tour kit component
-    - Update dependency `memoize-one` to ^6.0.0. #32936
-    - Update dependency `react-dates` to ^21.8.0. #32883
-    - Add Jetpack Changelogger
-    - Update dependency `@wordpress/hooks` to ^3.5.0
-    - Update dependency `@wordpress/icons` to ^8.1.0
-    - Add `className` prop for Pill component. #32605
-    - Removed unused react-router-dom dependency #33156
-    - Standardize lint scripts: add lint:fix
-    - Fix documentation for `TableCard` component
-    - Update `StepperProps` prop types. #32712 
```

### Other information:




### FOR PR REVIEWER ONLY:

-   [] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.
